File: v131016/TeeNew/Tool_AnnotationCallout.pas
unit Tool_AnnotationCallout;
{$I TeeDefs.inc}

interface

uses
  {$IFNDEF LINUX}
  Windows, Messages,
  {$ENDIF}
  SysUtils, Classes,
  {$IFDEF CLX}
  QGraphics, QControls, QForms, QDialogs, QExtCtrls, QStdCtrls, QComCtrls,
  {$ELSE}
  Graphics, Controls, Forms, Dialogs, ExtCtrls, StdCtrls, ComCtrls,
  {$ENDIF}
  Base, TeCanvas, TeePenDlg, TeeProcs, TeEngine, Chart, TeeTools, Series;

type
  TAnnotationCallout = class(TBaseForm)
    Button1: TButton;
    ButtonPen1: TButtonPen;
    Series1: TPointSeries;
    ChartTool1: TAnnotationTool;
    CheckBox1: TCheckBox;
    procedure Chart1MouseMove(Sender: TObject; Shift: TShiftState; X,
      Y: Integer);
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure CheckBox1Click(Sender: TObject);
    procedure Chart1ClickSeries(Sender: TCustomChart; Series: TChartSeries;
      ValueIndex: Integer; Button: TMouseButton; Shift: TShiftState; X,
      Y: Integer);
  private
    { Private declarations }
    Procedure SetCallout(AIndex:Integer);
  public
    { Public declarations }
  end;

implementation

{$IFNDEF CLX}
{$R *.dfm}
{$ELSE}
{$R *.xfm}
{$ENDIF}

uses EditChar;

// Returns Series point index that is nearest to xy position.
Function NearestPoint(ASeries:TChartSeries; x,y:Integer):Integer;
var Difference : Integer;
    tmpDif     : Integer;
    t          : Integer;
begin
  result:=-1;
  Difference:=-1;
  for t:=0 to ASeries.Count-1 do
  begin
    tmpDif:=Round(TeeDistance(ASeries.CalcXPos(t)-x,ASeries.CalcYPos(t)-y));

    if (Difference=-1) or (tmpDif<Difference) then
    begin
      Difference:=tmpDif;
      result:=t;
    end;
  end;
end;

procedure TAnnotationCallout.Chart1MouseMove(Sender: TObject;
  Shift: TShiftState; X, Y: Integer);
var tmp : Integer;
begin
  if CheckBox1.Checked then  // follow mouse
  begin
    // Locate nearest point to mouse...
    tmp:=NearestPoint(Series1, x, y);

    if tmp<>-1 then SetCallout(tmp);  // set annotation callout
  end;
end;

Procedure TAnnotationCallout.SetCallout(AIndex:Integer);
begin
  // Change annotation text
  ChartTool1.Text:='Point: '+IntToStr(AIndex)+#13+
                   'Value: '+Series1.ValueMarkText[AIndex];

  // Re-position annotation callout
  with ChartTool1.Callout do
  begin
    Visible:=True;
    XPosition:=Series1.CalcXPos(AIndex);
    YPosition:=Series1.CalcYPos(AIndex);
    ZPosition:=0;
  end;
end;

procedure TAnnotationCallout.FormCreate(Sender: TObject);
begin
  inherited;
  Series1.FillSampleValues;

  // force a first-time chart redrawing, to obtain valid
  // coordinates (Series1.CalcYPos, etc).
  Chart1.Draw;

  // Start positioning annotation callout at point index 5
  SetCallout(5);

  ChartTool1.Callout.Arrow.Visible:=True;

  ButtonPen1.LinkPen(ChartTool1.Callout.Arrow);
end;

procedure TAnnotationCallout.Button1Click(Sender: TObject);
begin
  EditChartTool(Self,ChartTool1);
end;

procedure TAnnotationCallout.CheckBox1Click(Sender: TObject);
begin
  Chart1.Title.Text.Clear;

  if CheckBox1.Checked then
  begin
    Chart1.Title.Text.Add('Move the mouse over points !');
    Series1.Cursor:=crDefault;
  end
  else
  begin
    Chart1.Title.Text.Add('Click a point !');
    Series1.Cursor:=crHandPoint;
  end;
end;

procedure TAnnotationCallout.Chart1ClickSeries(Sender: TCustomChart;
  Series: TChartSeries; ValueIndex: Integer; Button: TMouseButton;
  Shift: TShiftState; X, Y: Integer);
begin
  if not CheckBox1.Checked then
     SetCallout(ValueIndex);
end;

initialization
  RegisterClass(TAnnotationCallout);
end.

File: TeeNew/Builder2009/Tool_AnnotationCallout.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Tool_AnnotationCallout.h"
#include "EditChar.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Base"
#pragma link "TeePenDlg"
#pragma link "TeeTools"
#pragma resource "*.dfm"
TAnnotationCO *AnnotationCO;
//---------------------------------------------------------------------------
__fastcall TAnnotationCO::TAnnotationCO(TComponent* Owner)
  : TBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TAnnotationCO::SetCallout(int AIndex)
{
  // Change annotation text
  ChartTool1->Text = "Point: "+IntToStr(AIndex)+TeeLineSeparator+
                   "Value: "+Series1->ValueMarkText[AIndex];

  // Re-position annotation callout
  ChartTool1->Callout->Visible = true;
  ChartTool1->Callout->XPosition = Series1->CalcXPos(AIndex);
  ChartTool1->Callout->YPosition = Series1->CalcYPos(AIndex);
  ChartTool1->Callout->ZPosition = 0;
}
//---------------------------------------------------------------------------
// Returns Series point index that is nearest to xy position.
int __fastcall TAnnotationCO::NearestPoint(TChartSeries *ASeries, int x, int y)
{
  int res = -1;
  int Difference = -1;
  int tmpDif;
  for (int t=0; t<ASeries->Count(); t++)
  {
    tmpDif = int(TeeDistance(ASeries->CalcXPos(t)-x,ASeries->CalcYPos(t)-y));
    if (Difference==-1 || tmpDif<Difference)
    {
      Difference = tmpDif;
      res = t;
    }
  }
  return res;
}
//---------------------------------------------------------------------------

void __fastcall TAnnotationCO::Chart1MouseMove(TObject *Sender,
      TShiftState Shift, int X, int Y)
{
  if (CheckBox1->Checked) // follow mouse
  {
    // Locate nearest point to mouse...
    int tmp = NearestPoint(Series1, X, Y);
    if (tmp!=-1) SetCallout(tmp);  // set annotation callout
  }
}
//---------------------------------------------------------------------------

void __fastcall TAnnotationCO::FormCreate(TObject *Sender)
{
  Series1->FillSampleValues(10);

  // force a first-time chart redrawing, to obtain valid
  // coordinates (Series1->CalcYPos, etc).
  Chart1->Draw();

  // Start positioning annotation callout at point index 5
  SetCallout(5);

  ChartTool1->Callout->Arrow->Visible = true;

  ButtonPen1->LinkPen(ChartTool1->Callout->Arrow);
}
//---------------------------------------------------------------------------

void __fastcall TAnnotationCO::Button1Click(TObject *Sender)
{
  EditChartTool(this,ChartTool1);
}
//---------------------------------------------------------------------------

void __fastcall TAnnotationCO::CheckBox1Click(TObject *Sender)
{
  Chart1->Title->Text->Clear();

  if (CheckBox1->Checked)
  {
    Chart1->Title->Text->Add("Move the mouse over points !");
    Series1->Cursor = crDefault;
  }
  else
  {
    Chart1->Title->Text->Add("Click a point !");
    Series1->Cursor = crHandPoint;
  }
}
//---------------------------------------------------------------------------

void __fastcall TAnnotationCO::Chart1ClickSeries(TCustomChart *Sender,
      TChartSeries *Series, int ValueIndex, TMouseButton Button,
      TShiftState Shift, int X, int Y)
{
  if (!CheckBox1->Checked) SetCallout(ValueIndex);
}
//---------------------------------------------------------------------------

